File: Divide and Conquer/quicksort.py
# Here
#     unsorted_arr  = The unsorted array
#     p             = First index of array
#     r             = Last index of array

def partation(arr, p, r): 

    x = arr[r]
    i = p - 1
    for j in range(p, r):
        if arr[j] <= x:
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]     # Swapping the valuess
    
    arr[i+1], arr[r] = arr[r], arr[i+1]         # Swapping the valuess
    return i + 1



def quicksort(arr, p, r):
    if p < r:
        q  = partation(arr, p, r)
        quicksort(arr, p, q - 1)
        quicksort(arr, p + 1, r)
    
    return arr


my_list = [5, 8, 7, 4, 2, 9, 12, 13, 11, 23, 1, -1, -23, 34, 545]
n = len(my_list)

quicksort(my_list, 0, n - 1)
print(my_list)

File: Javascript/Divide and Conquer/quickSort.js
// Define the function for partitioning the array
function partition(arr, p, r) {
    let x = arr[r];
    let i = p - 1;

    for (let j = p; j < r; j++) {
        if (arr[j] <= x) {
            i++;
            [arr[i], arr[j]] = [arr[j], arr[i]];  // Swapping the values
        }
    }

    [arr[i+1], arr[r]] = [arr[r], arr[i+1]];  // Swapping the values
    return i + 1;
}

// Define the function for performing quicksort
function quicksort(arr, p, r) {
    if (p < r) {
        let q = partition(arr, p, r);
        quicksort(arr, p, q - 1);
        quicksort(arr, q + 1, r);
    }

    return arr;
}

// Define the main function
function main() {
    let my_list = [5, 8, 7, 4, 2, 9, 12, 13, 11, 23, 1, -1, -23, 34, 545];
    let n = my_list.length;

    quicksort(my_list, 0, n - 1);
    console.log(my_list);
}

// Call the main function
main();
